File: crates/ferrolex-python/bench/lookup.py
import json
import sys
import time

sys.path.insert(0, sys.argv[1])
import ferrolex_python

words = ["word" + str(index) for index in range(4096)] + ["ferrolex"]
queries = [
    "word" + str(index % 4096) if index % 3 == 0 else "missing" + str(index)
    for index in range(12000)
]
checker = ferrolex_python.Checker("\n".join(words))
baseline = set(words)


def measure(check):
    recognized = 0
    start = time.perf_counter_ns()
    for query in queries:
        if check(query):
            recognized += 1
    return time.perf_counter_ns() - start, recognized


native_ns, native_recognized = measure(checker.check)
baseline_ns, baseline_recognized = measure(lambda query: query in baseline)
if native_recognized != baseline_recognized:
    raise RuntimeError("binding and set baseline disagree on recognized queries")
if "ferrolex" not in checker.suggest("ferolex"):
    raise RuntimeError("binding did not expose the expected suggestion")

print(
    json.dumps(
        {
            "binding": "python-pyo3",
            "queries": len(queries),
            "nativeNs": native_ns,
            "setBaselineNs": baseline_ns,
            "recognized": native_recognized,
        }
    )
)

File: crates/ferrolex-node/bench/lookup.js
'use strict';

const native = require(process.argv[2]);
const words = Array.from({ length: 4096 }, (_, index) => 'word' + index);
words.push('ferrolex');
const queries = Array.from({ length: 12000 }, (_, index) =>
  index % 3 === 0 ? 'word' + (index % 4096) : 'missing' + index,
);

const checker = new native.Checker(words.join('\n'));
const baseline = new Set(words);

function measure(check) {
  let recognized = 0;
  const start = process.hrtime.bigint();
  for (const query of queries) {
    if (check(query)) {
      recognized += 1;
    }
  }
  return { elapsedNs: Number(process.hrtime.bigint() - start), recognized };
}

const nativeResult = measure((query) => checker.check(query));
const baselineResult = measure((query) => baseline.has(query));
if (nativeResult.recognized !== baselineResult.recognized) {
  throw new Error('binding and Set baseline disagree on recognized queries');
}
if (!checker.suggest('ferolex').includes('ferrolex')) {
  throw new Error('binding did not expose the expected suggestion');
}

console.log(JSON.stringify({
  binding: 'node-napi',
  queries: queries.length,
  nativeNs: nativeResult.elapsedNs,
  setBaselineNs: baselineResult.elapsedNs,
  recognized: nativeResult.recognized,
}));
